### ISSUE TITLE
feature: Remove Component from ValidationState, add IValidationState

### ISSUE THREAD

**Issue by thargy** [reactiveui/ReactiveUI.Validation#123]
**Is your feature request related to a problem? Please describe.**
At the core of ReactiveUI.Validation are `IValidationComponent`s that expose `IObservable<ValidationState>`.  All the `ValidationRule` extensions fundamentally boil down to producing a `ValidationState` when necessary.  However, it is not an ideal extensibility point because `ValidationState` is a sealed class, which includes a back-reference to the 'owner' Component.  

The ability to produce an `ValidationState`s directly would be of great benefit.  Even better would be for the API to accept an interface `IValidationState` which _does not require_ the back reference to the owner  `IValidationComponent`.  This would allow consumption of observables that implement `IValidationState`, maximising efficiency, as compared to using one of the existing `ValidationRule` extension methods to manufacture the `ValidationState` class for each change in validation (see for example the new method added in #116, which requires two functions to extract the validity and text from a TState).

**Describe the solution you'd like**
I suggest that a new `IValidationState` interface is created:
```csharp
    public interface IValidationState
    {
        bool IsValid { get; }
        ValidationText Text { get; }
    }
```

This interface should be used wherever `ValidationState` is currently used in the library.

`ValidationState` should implement `IValidationState`, but the `Component` property should be removed.  This would then be a utility struct for consumers to easily provide a default implementation of `IValidationState`.  There is also a strong case for changing `ValidationState` into a struct - it being a single pointer and a bool, and having a sensible default state (though care should be taken to handle a `null` Text); alternatively, the class should not be sealed, to allow it to be used as a base class for implementors.

A new ValidationRule extension should be added that that accepts an `IObservable<IValidationState>`.

**Describe alternatives you've considered**
This is not a _required_ feature, so no alternatives are required, however, it should be considered that accepting an interface will allow for consumers to supply simple Observables of any struct/class that implements that interface, and not use one of the more complex overloads.

Further, the `Compoment` property appearing in the child element is code smell as there is a forward and back reference between `Component` and `ValidationState`, this is generally considered bad practice as it allows for data to become inconsistent, which can introduce bugs.

**Describe suggestions on how to achieve the feature**
A review of `ValidationState.Component` shows that the only _meaningful_ usage is in [ReactiveValidationObject.OnValidationStatusChange](https://github.com/reactiveui/ReactiveUI.Validation/blob/7a38594a8634fd11600528aea4cdb576281e7ca8/src/ReactiveUI.Validation/Helpers/ReactiveValidationObject.cs#L130), a private method call, called from an observable in the constructor:
https://github.com/reactiveui/ReactiveUI.Validation/blob/7a38594a8634fd11600528aea4cdb576281e7ca8/src/ReactiveUI.Validation/Helpers/ReactiveValidationObject.cs#L51-L68

This can be rewritten to pass the `IValidationCompoment` to the method so that it does not need to get it from the `IValidationState`.

**Additional context**
Example usage:
```csharp
// IObservable<ValidationState> where ValidationState implements IValidationState
var passwordsObservable =
    this.WhenAnyValue(
        x => x.Password,
        x => x.ConfirmPassword,
        (password, confirmation) => new ValidationState(
				state.Password == state.Confirmation,
				$"Passwords must match: {state.Password} != {state.Confirmation}"));

this.ValidationRule(
    vm => vm.ConfirmPassword,
    passwordsObservable);

```

As noted, this is an improvement over the existing example, in that it only produces one ValidationState object, and does not need to make two function calls on each change.  It is also very extensible as any object that implements `IValidationState` can be consumed.

**Comment by worldbeater** [comment]
Thanks for describing your vision of the modified API in detail! Sounds really reasonable. Worth noting btw that we should define the new interface as somewhat like the following to avoid introducing a massive breaking change:

```cs
public interface IValidationState
{
    bool IsValid { get; }
    ValidationText Text { get; }
    [Obsolete("This property will be removed soon.")]
    IValidationComponent Component { get; }
}
```

Currently, we are planning to give people a couple of months to adapt before removing the `[Obsolete]` APIs completely.

**Comment by thargy** [comment]
Hi @worldbeater,

As `IValidationState` would be a new interface anyway it does not need (and shouldn't have) the `Component` property.  It won't break any existing code for it to be absent as no existing code would know about the interface.

Changing any API method that acccepts `ValidationState` to now accept `IValidationState` should likewise be low impact as they would accept the existing `ValidationState` object (so long as it is modified to implement `IValidationState`.  Of course, such code would not be able to use the `Component` property, as it doesn't exist on the new interface (it's only used once in the library anyway as mentioned).

To make changes to `ValidationState` backwards compatible you can make the `Component` nullable, and obsolete it.  You can also change the constructor parameter for the component to be optional and obsolete and remove the null check.  If you want to belt and brace it, you could make the whole object obsolete, and instead provide a new ValidationState struct in a different namespace location, that doesn't include Component.

**Comment by worldbeater** [comment]
Assuming we have `IValidationComponent.ValidationStatusChange` exposed to the outer world, and also assuming the fact, that the `ValidationContext` class implements `IValidationComponent`, changing `ValidationStatusChange` from `IObservable<ValidationState>` to `IObservable<IValidationState>` would break existing code like shown in this snippet:

```cs
IValidationComponent context = new ValidationContext();
context.ValidationStatusChange
    .Select(state => state.Component) // Who knows, there may be folks who rely on this Component property.
```

Adding an `[Obsolete] Component` property to the new interface could make the transition less breaking and less painful. I mean, if we are introducing a new `IValidationState`, we are planning to change the return type of `ValidationStatusChange` observable anyway. Or do you mean it is worth continuing returning `IObservable<ValidationState>`, and change the return type to `IObservable<IValidationState>` later, when we will be removing stuff marked as `[Obsolete]`?

**Comment by thargy** [comment]
That's a good point, though I wonder how many people actually do use the `Component`.  My suspicion is it would be very rare at best as the existing recommended way to use the library is through the `ValidationRule` extension overloads and the underlying components and states are not particularly surfaced.  The problem of adding it to the new interface is that people coming to use it will have to implement it (though you could provide a default implementation in the interface that returns `null` to prevent the need for its explicit implementation), whereas, the goal would be to discourage its use entirely.  It seems a shame to add an Obsolete default implementation property to a brand new interface, but it is a safe option.

**Comment by thargy** [comment]
This has been fully addressed in #130.
